libs/shared/src/store/immerReducer.ts

```diff
@@ -1,68 +1,180 @@
-import { ImmerReducer,  createReducerFunction, createActionCreators, ImmerReducerClass, ImmerReducerState } from "immer-reducer";
-import { put } from "redux-saga/effects";
-import { Reducer } from "redux";
+import { ImmerReducer, createReducerFunction, createActionCreators, ImmerReducerClass, ImmerReducerState } from "immer-reducer";
+import { put, select, SelectEffect, takeLatest } from "redux-saga/effects";
+import { Action, bindActionCreators, Reducer } from "redux";
+import { useDispatch, useSelector } from 'react-redux';
+import { useInjectReducer, useInjectSaga } from 'redux-injectors';
+import { Saga } from 'redux-saga';
 
-interface Action {
-  type: string;
+
+//
+// Duplicating the ActionsType from ImmerReducer.  This
+// is unfortunately necessary because the package version
+// requires the input to be a class (and we can only provide an interface)
+type FirstOrAll<T> = T extends [infer V] ? V : T;
+type ArgumentsType<T> = T extends (...args: infer V) => any ? V : never;
+export interface ImmerActionCreator<ActionTypeType, Payload extends any[]> {
+  readonly type: ActionTypeType;
+  (...args: Payload): {
+      type: ActionTypeType;
+      payload: FirstOrAll<Payload>;
+  };
+}
+
+type ActionsType<Interface> = {
+  [K in keyof Interface] : ImmerActionCreator<K, ArgumentsType<Interface[K]>>
 }
 
-export class TheCoinReducer<T> extends ImmerReducer<T> {
+export function BaseReducer<U, T>(key: string, initialState: T) {
+  // Return new class so we can re-use the static declarations
+  const BaseReducer = class BaseReducer extends ImmerReducer<T> {
+
+    static _reducer: any;
+    static _actions: any;
```

Comment:
yes, `any` - but the properly-typed vesrions are below, this is just to get around the fact that we don't actually have proper definitions for these types yet